substrate/client/src/client.rs

```diff
@@ -261,9 +303,21 @@ impl<B, E> Client<B, E> where
 
 		let is_new_best = header.number == self.backend.blockchain().info()?.best_number + 1;
 		trace!("Imported {}, (#{}), best={}", block::HeaderHash::from(header.blake2_256()), header.number, is_new_best);
-		transaction.set_block_data(header, body, Some(justification.uncheck().into()), is_new_best)?;
+		transaction.set_block_data(header.clone(), body, Some(justification.uncheck().into()), is_new_best)?;
 		transaction.set_storage(overlay.drain())?;
 		self.backend.commit_operation(transaction)?;
+
+		if origin == BlockOrigin::NetworkBroadcast || origin == BlockOrigin::Own {
+			let notification = BlockImportNotification {
+				origin: origin,
+				header: header,
+				is_new_best: is_new_best,
+			};
+			if let Err(e) = self.import_notification_sink.lock().try_send(notification) {
```

Comment:
this fails if the buffer is full; seems dangerous to discard the notification in this case. maybe use the `Sink::send` api and wait for the future to resolve?

Reply:
Import is more important and should not block on the notification handler. Normally this should never happen but if it does, skipping a notification is fine since there will be a new block eventually anyway. The handlers are not supposed to expect to receive notifications for every block anyway. It is quite possible that the node goes to sleep or loses connectivity for a while and when it goes back online there won't be any notifications for all the blocks that arrived in the meantime.

Reply:
> skipping a notification is fine since there will be a new block eventually anyway

there won't be a new block if not enough authorities start the agreement process (e.g. if there are 1/3 offline and one of the "honest" nodes skips this notification)

Reply:
Increased the buffer to 65536 for now. Unfortunately it is preallocated.

Reply:
still dangerous but probably workable for now. we should write a fix after this PR

Reply:
I'll address that in the following PR. Probably ny replacing `multiqueue` with a vector of mpsc channels.